### PR TITLE
Ban classes conforming to _ObjectiveCBridgeable

### PR DESCRIPTION
**Pull request by beccadax**
This PR adds a special case to the declaration checker which makes it reject attempts to conform classes to `_ObjectiveCBridgeable`. It doesn’t make much sense to bridge a class to a different class when it comes into Swift, and PrintAsObjC has never handled these conformances correctly.

Banning is an aggressive solution, but we don’t promise source compatibility with underscored declarations, so it’s a matter of how badly this breaks things in practice. If this causes too much pain, we can instead add appropriate test coverage and fix PrintAsObjC.

This PR also changes a "can't happen" code path in PrintAsObjC which used to print syntactically invalid code into the generated header to instead abort with a detailed message asking the user to file a bug. Classes conforming to `_ObjectiveCBridgeable` used to end up in this place; if anything else does, I want to know about it.

Fixes rdar://62890964.

### CODE REVIEW

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please test source compatibility

**Comment by brentdax** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
@swift-ci please test source compatibility

**Comment by brentdax** [comment]
The only source compatibility suite failure looks spurious.

File: lib/Sema/TypeCheckProtocol.cpp [hunk 1]
```diff
@@ -1742,13 +1742,27 @@ checkIndividualConformance(NormalProtocolConformance *conformance,
...
   }
 
-  // If the protocol requires a class, non-classes are a non-starter.
-  if (Proto->requiresClass() && !canT->getClassOrBoundGenericClass()) {
-    C.Diags.diagnose(ComplainLoc,
-                     diag::non_class_cannot_conform_to_class_protocol, T,
-                     ProtoType);
-    conformance->setInvalid();
-    return conformance;
+  if (canT->getClassOrBoundGenericClass()) {
```
Comment (slavapestov) [inline lib/Sema/TypeCheckProtocol.cpp:1745]: Can you put this next to the existing `ObjectiveCBridgeable` special case in ConformanceChecker::checkConformance()?

**Comment by shahmishal** [comment]
Please update the base branch to `main` by Oct 5th otherwise the pull request will be closed automatically. 

* How to change the base branch: ([Link](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-base-branch-of-a-pull-request))
* More detail about the branch update: ([Link](https://forums.swift.org/t/updating-branch-names/40412/))

**Pull request closed without merging**